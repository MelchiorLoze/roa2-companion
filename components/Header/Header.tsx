import FontAwesome from '@expo/vector-icons/FontAwesome';
import { NativeStackHeaderProps } from '@react-navigation/native-stack';
import { Image } from 'expo-image';
import { StyleSheet, Text, View } from 'react-native';

import { useAuth } from '@/contexts/AuthContext/AuthContext';
import { useCurrencyBalance } from '@/hooks/business';

export const Header = ({ options }: NativeStackHeaderProps) => {
  const { logout } = useAuth();
  const { coinsBalance, bucksBalance } = useCurrencyBalance();

  return (
    <>
      <View style={styles.topContainer}>
        <Text>{options.title}</Text>
        <FontAwesome.Button
          backgroundColor="transparent"
          color="red"
          iconStyle={styles.logoutIcon}
          name="power-off"
          onPress={logout}
        />
      </View>
      <View style={styles.bottomContainer}>
        <View style={styles.currencyContainer}>
          <Image contentFit="contain" source={require('@/assets/images/coins.png')} style={styles.currencyIcon} />
          <Text style={styles.currencyLabel}>{coinsBalance}</Text>
        </View>
        <View style={styles.currencyContainer}>
          <Image contentFit="contain" source={require('@/assets/images/bucks.png')} style={styles.currencyIcon} />
          <Text style={styles.currencyLabel}>{bucksBalance}</Text>
        </View>
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  topContainer: {
    position: 'relative',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: 8,
    backgroundColor: 'white',
  },
  bottomContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    padding: 4,
    backgroundColor: 'darkblue',
  },
  currencyContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 4,
  },
  currencyLabel: {
    color: 'white',
  },
  currencyIcon: {
    width: 24,
    height: 24,
  },
  logoutIcon: {
    marginRight: 0,
  },
});
