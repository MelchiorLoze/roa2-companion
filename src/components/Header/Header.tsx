import { useRouter } from 'expo-router';
import { Text, View } from 'react-native';
import { StyleSheet } from 'react-native-unistyles';

import { IconButton } from '../IconButton/IconButton';
import { Separator } from '../Separator/Separator';
import { CurrenciesBalance } from './CurrenciesBalance/CurrenciesBalance';

type Props = {
  title?: string;
  showCurrencies?: boolean;
  withBackNavigation?: boolean;
};

export const Header = ({ title, showCurrencies, withBackNavigation }: Readonly<Props>) => {
  const router = useRouter();

  return (
    <>
      {showCurrencies ? <CurrenciesBalance style={styles.topContainer} /> : <View style={styles.topSpacing} />}
      <Separator height={2} variant="accent" />
      {title && (
        <View style={styles.bottomContainer(withBackNavigation)}>
          {withBackNavigation && (
            <IconButton iconName="arrow-back" onPress={router.back} size={24} style={styles.backButton} />
          )}
          <Text style={styles.title}>{title}</Text>
        </View>
      )}
    </>
  );
};

const styles = StyleSheet.create((theme, runtime) => ({
  topSpacing: {
    height: theme.spacing.s + runtime.insets.top,
    backgroundColor: theme.color.highlight,
  },
  topContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    padding: theme.spacing.s,
    paddingTop: theme.spacing.s + runtime.insets.top,
    backgroundColor: theme.color.highlight,
  },
  bottomContainer: (withBackNavigation?: boolean) => ({
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: withBackNavigation ? theme.spacing.s : theme.spacing.m,
    paddingHorizontal: withBackNavigation ? theme.spacing.s : theme.spacing.l,
    backgroundColor: theme.color.background,
  }),
  backButton: {
    padding: theme.spacing.s,
  },
  title: {
    width: '100%',
    fontFamily: theme.font.primary.italic,
    fontSize: 24,
    color: theme.color.white,
    textTransform: 'uppercase',
  },
}));
