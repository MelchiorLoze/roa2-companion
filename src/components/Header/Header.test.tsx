import { fireEvent, render, screen } from '@testing-library/react-native';
import { useRouter } from 'expo-router';

import { Header } from './Header';

jest.mock('expo-router');
const backMock = jest.fn();
jest.mocked(useRouter).mockReturnValue({
  back: backMock,
} as unknown as ReturnType<typeof useRouter>);

jest.mock('@/hooks/business/useCurrencyBalance/useCurrencyBalance', () => ({
  useCurrencyBalance: jest.fn(() => ({
    coinsBalance: 0,
    bucksBalance: 0,
    medalsBalance: 0,
    isLoading: false,
    isError: false,
  })),
}));

describe('Header', () => {
  it('renders without crashing when no props are provided', () => {
    render(<Header />);

    expect(screen.queryByTestId('currencies-balance')).toBeNull();
    expect(screen.queryByText('Test Title')).toBeNull();
    expect(screen.queryByRole('button', { name: 'Back' })).toBeNull();
  });

  it('renders with currencies balance', () => {
    render(<Header showCurrencies />);

    screen.getByTestId('currencies-balance');
    expect(screen.queryByText('Test Title')).toBeNull();
    expect(screen.queryByRole('button', { name: 'Back' })).toBeNull();
  });

  it('renders with title', () => {
    render(<Header title="Test Title" />);

    expect(screen.queryByTestId('currencies-balance')).toBeNull();
    screen.getByText('Test Title');
    expect(screen.queryByRole('button', { name: 'Back' })).toBeNull();
  });

  it('renders with title and back navigation', () => {
    render(<Header title="Test Title" withBackNavigation />);

    expect(screen.queryByTestId('currencies-balance')).toBeNull();
    screen.getByText('Test Title');

    const backButton = screen.getByRole('button');
    fireEvent.press(backButton);

    expect(backMock).toHaveBeenCalledTimes(1);
  });

  it('does not render back navigation when no title is provided', () => {
    render(<Header withBackNavigation />);

    expect(screen.queryByTestId('currencies-balance')).toBeNull();
    expect(screen.queryByText('Test Title')).toBeNull();
    expect(screen.queryByRole('button')).toBeNull();
  });

  it('matches the snapshot', () => {
    const tree = render(<Header showCurrencies title="Test title" withBackNavigation />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
