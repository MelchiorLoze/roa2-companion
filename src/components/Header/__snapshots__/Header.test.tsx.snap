// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches the snapshot 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2B1F74",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "padding": 8,
        "paddingTop": 8,
      }
    }
    testID="currencies-balance"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <ViewManagerAdapter_ExpoImage
        containerViewRef={"[React.ref]"}
        contentFit="contain"
        contentPosition={
          {
            "left": "50%",
            "top": "50%",
          }
        }
        height={20}
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={[]}
        source={
          [
            {
              "testUri": "../../../src/assets/images/currency/coins.png",
            },
          ]
        }
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
        transition={null}
        width={20}
      />
      <Text
        style={
          {
            "color": "white",
            "fontFamily": "AgencyFB-Black",
            "fontSize": 16,
          }
        }
      >
        <Text
          style={
            {
              "color": "#070032",
            }
          }
        >
          0000000
        </Text>
        0
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <ViewManagerAdapter_ExpoImage
        containerViewRef={"[React.ref]"}
        contentFit="contain"
        contentPosition={
          {
            "left": "50%",
            "top": "50%",
          }
        }
        height={20}
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={[]}
        source={
          [
            {
              "testUri": "../../../src/assets/images/currency/bucks.png",
            },
          ]
        }
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
        transition={null}
        width={20}
      />
      <Text
        style={
          {
            "color": "white",
            "fontFamily": "AgencyFB-Black",
            "fontSize": 16,
          }
        }
      >
        <Text
          style={
            {
              "color": "#070032",
            }
          }
        >
          0000000
        </Text>
        0
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <ViewManagerAdapter_ExpoImage
        containerViewRef={"[React.ref]"}
        contentFit="contain"
        contentPosition={
          {
            "left": "50%",
            "top": "50%",
          }
        }
        height={20}
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={[]}
        source={
          [
            {
              "testUri": "../../../src/assets/images/currency/medals.png",
            },
          ]
        }
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
        transition={null}
        width={20}
      />
      <Text
        style={
          {
            "color": "white",
            "fontFamily": "AgencyFB-Black",
            "fontSize": 16,
          }
        }
      >
        <Text
          style={
            {
              "color": "#070032",
            }
          }
        >
          0000000
        </Text>
        0
      </Text>
    </View>
  </View>,
  <View
    style={
      {
        "backgroundColor": "#FEDD84",
        "height": 2,
      }
    }
  />,
  <ViewManagerAdapter_ExpoLinearGradient
    colors={
      [
        4278518042,
        4280687939,
      ]
    }
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "paddingHorizontal": 8,
        "paddingVertical": 8,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      role="button"
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "justifyContent": "center",
          },
          {
            "padding": 8,
          },
        ]
      }
    >
      <
        color="white"
        name="arrow-back"
        size={24}
      />
    </View>
    <Text
      style={
        {
          "color": "white",
          "fontFamily": "FranklinGothicDemiCond-Italic",
          "fontSize": 24,
          "textTransform": "uppercase",
          "width": "100%",
        }
      }
    >
      Test title
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>,
]
`;
